File: cs/Challenge/IncorrectImplementations/DoNotOpen.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Challenge.Infrastructure;

namespace Challenge.IncorrectImplementations
{
    #region Не подглядывать!

    [IncorrectImplementation]
    public class WordsStatisticsL2 : WordsStatistics
    {
        public override void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (string.IsNullOrWhiteSpace(word)) return;
            int count;
			statistics[word.ToLower()] = 1 + (statistics.TryGetValue(word.ToLower(), out count) ? count : 0);
		}
    }

    [IncorrectImplementation]
    public class WordsStatisticsL3 : WordsStatistics
    {
        public override void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (string.IsNullOrWhiteSpace(word)) return;
            if (word.Length > 10) word = word.Substring(0, 10);
            else if (word.Length > 5) word = word.Substring(0, word.Length - 2);
            int count;
			statistics[word.ToLower()] = 1 + (statistics.TryGetValue(word.ToLower(), out count) ? count : 0);
		}
    }

    [IncorrectImplementation]
    public class WordsStatisticsL4 : WordsStatistics
    {
        public override void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (string.IsNullOrWhiteSpace(word)) return;
            if (word.Length - 1 > 10) word = word.Substring(0, 10);
            int count;
			statistics[word.ToLower()] = 1 + (statistics.TryGetValue(word.ToLower(), out count) ? count : 0);
		}
    }

    [IncorrectImplementation]
    public class WordsStatisticsC : WordsStatistics
    {
        public override void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (string.IsNullOrWhiteSpace(word)) return;
            if (word.Length > 10) word = word.Substring(0, 10);
            if (!statistics.ContainsKey(word.ToLower()))
                statistics[word] = 0;
            statistics[word]++;
        }
    }

    [IncorrectImplementation]
    public class WordsStatisticsE : WordsStatistics
    {
        public override void AddWord(string word)
        {
            if (string.IsNullOrWhiteSpace(word)) throw new ArgumentNullException(nameof(word));
            if (word.Length > 10) word = word.Substring(0, 10);
            int count;
			statistics[word.ToLower()] = 1 + (statistics.TryGetValue(word.ToLower(), out count) ? count : 0);
		}
    }

    [IncorrectImplementation]
    public class WordsStatisticsE2 : WordsStatistics
    {
        public override void AddWord(string word)
        {
            if (string.IsNullOrWhiteSpace(word)) return;
            if (word.Length > 10) word = word.Substring(0, 10);
            int count;
			statistics[word.ToLower()] = 1 + (statistics.TryGetValue(word.ToLower(), out count) ? count : 0);
		}
    }

    [IncorrectImplementation]
    public class WordsStatisticsE3 : WordsStatistics
    {
        public override void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (word.Length > 10) word = word.Substring(0, 10);
            if (string.IsNullOrWhiteSpace(word)) return;
            int count;
			statistics[word.ToLower()] = 1 + (statistics.TryGetValue(word.ToLower(), out count) ? count : 0);
		}
    }

    [IncorrectImplementation]
    public class WordsStatisticsE4 : WordsStatistics
    {
        public override void AddWord(string word)
        {
            if (word.Length == 0 || word.All(char.IsWhiteSpace)) return;
            if (word.Length > 10) word = word.Substring(0, 10);
            int count;
			statistics[word.ToLower()] = 1 + (statistics.TryGetValue(word.ToLower(), out count) ? count : 0);
		}
    }

    [IncorrectImplementation]
    public class WordsStatisticsO1 : WordsStatistics
    {
        public override IEnumerable<WordCount> GetStatistics()
        {
            return statistics
                .Select(WordCount.Create)
                .OrderBy(wordCount => wordCount.Word);
        }
    }

    [IncorrectImplementation]
    public class WordsStatisticsO2 : WordsStatistics
    {
        public override IEnumerable<WordCount> GetStatistics()
        {
            return statistics
                .Select(WordCount.Create)
                .OrderBy(wordCount => wordCount.Count);
        }
    }

    [IncorrectImplementation]
    public class WordsStatisticsO3 : WordsStatistics
    {
        public override IEnumerable<WordCount> GetStatistics()
        {
            return base.GetStatistics().OrderBy(wordCount => wordCount.Word);
        }
    }

    [IncorrectImplementation]
    public class WordsStatisticsO4 : WordsStatistics
    {
        public override IEnumerable<WordCount> GetStatistics()
        {
            return base.GetStatistics()
                .OrderByDescending(wordCount => wordCount.Count)
                .ThenByDescending(wordCount => wordCount.Word);
        }
    }

    [IncorrectImplementation]
    public class WordsStatisticsCR : IWordsStatistics
    {
        private readonly IDictionary<string, int> statistics =
            new Dictionary<string, int>();

        public void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (string.IsNullOrEmpty(word)) return;
            if (word.Length > 10) word = word.Substring(0, 10);
            word = word.ToLower();
            int count;
            statistics[word] = 1 + (statistics.TryGetValue(word, out count) ? count : 0);
        }

        public IEnumerable<WordCount> GetStatistics()
        {
            return statistics
                .Select(WordCount.Create)
                .OrderByDescending(wordCount => wordCount.Count)
                .ThenBy(wordCount => wordCount.Word);
        }
    }

    [IncorrectImplementation]
    public class WordsStatisticsSTA : IWordsStatistics
    {
        private static readonly IDictionary<string, int> statistics = new Dictionary<string, int>();

        public WordsStatisticsSTA()
        {
	        statistics.Clear();
        }

        public void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (string.IsNullOrWhiteSpace(word)) return;
            if (word.Length > 10) word = word.Substring(0, 10);
            int count;
			statistics[word.ToLower()] = 1 + (statistics.TryGetValue(word.ToLower(), out count) ? count : 0);
		}

        public IEnumerable<WordCount> GetStatistics()
        {
            return statistics
                .Select(WordCount.Create)
                .OrderByDescending(wordCount => wordCount.Count)
                .ThenBy(wordCount => wordCount.Word);
        }
    }

    [IncorrectImplementation]
    public class WordsStatistics123 : IWordsStatistics
    {
        private const int MAX_SIZE = 1237;

        private readonly int[] statistics = new int[MAX_SIZE];
        private readonly string[] words = new string[MAX_SIZE];

        public void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (string.IsNullOrWhiteSpace(word)) return;
            if (word.Length > 10) word = word.Substring(0, 10);
            var index = Math.Abs(word.ToLower().GetHashCode()) % MAX_SIZE;
	        statistics[index]++;
            words[index] = word.ToLower();
        }

        public IEnumerable<WordCount> GetStatistics()
        {
            return statistics.Zip(words, (s, w) => new WordCount(w, s))
                .Where(t => t.Count > 0)
                .OrderByDescending(t => t.Count)
                .ThenBy(t => t.Word);
        }
    }

    [IncorrectImplementation]
    public class WordsStatisticsQWE : IWordsStatistics
    {
        private readonly IDictionary<string, int> statistics = new Dictionary<string, int>();

        public void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (string.IsNullOrWhiteSpace(word)) return;
            if (word.Length > 10) word = word.Substring(0, 10);
            word = ToLower(word);
            int count;
	        statistics[word] = 1 + (statistics.TryGetValue(word, out count) ? count : 0);
        }

        public IEnumerable<WordCount> GetStatistics()
        {
            return statistics
                .Select(WordCount.Create)
                .OrderByDescending(wordCount => wordCount.Count)
                .ThenBy(wordCount => wordCount.Word);
        }

        private char ToLower(char c)
        {
            if ("QWERTYUIOPLJKHGFDSAZXCVBNM".Contains(c))
                return (char)(c - 'D' + 'd');
            else if ("ЙЦУКЕНГШЩЗФЫВАПРОЛДЯЧСМИТЬ".Contains(c))
                return (char)(c - 'Я' + 'я');
            return c;
        }

        private string ToLower(string s)
        {
            return new string(s.Select(ToLower).ToArray());
        }
    }

    [IncorrectImplementation]
    public class WordsStatistics998 : IWordsStatistics
    {
        private readonly List<WordCount> statistics = new List<WordCount>();

        public void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (string.IsNullOrWhiteSpace(word)) return;
            if (word.Length > 10) word = word.Substring(0, 10);
	        var lowerCaseWord = word.ToLower();
	        var wordCount = statistics.FirstOrDefault(s => s.Word == lowerCaseWord);
            if (wordCount.Word != null)
                statistics.Remove(wordCount);
            else
                wordCount = new WordCount(lowerCaseWord, 0);
            statistics.Add(new WordCount(wordCount.Word, wordCount.Count - 1));

            statistics.Sort((a, b) => a.Count == b.Count
                ? string.Compare(a.Word, b.Word, StringComparison.Ordinal)
                : a.Count - b.Count);
        }

        public IEnumerable<WordCount> GetStatistics()
        {
            return statistics.Select(w => new WordCount(w.Word, -w.Count));
        }
    }

    [IncorrectImplementation]
    public class WordsStatistics999 : IWordsStatistics
    {
        private readonly HashSet<string> usedWords = new HashSet<string>();
        private readonly List<WordCount> statistics = new List<WordCount>();

        public void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (string.IsNullOrWhiteSpace(word)) return;
            if (word.Length > 10) word = word.Substring(0, 10);
            word = word.ToLower();
            if (usedWords.Contains(word))
            {
                var stat = statistics.First(s => s.Word == word);
                statistics.Remove(stat);
                statistics.Add(new WordCount(stat.Word, stat.Count + 1));
            }
            else
            {
                statistics.Add(new WordCount(word, 1));
                usedWords.Add(word);
            }
        }

	    public IEnumerable<WordCount> GetStatistics()
	    {
		    return statistics
                .OrderByDescending(t => t.Count)
			    .ThenBy(t => t.Word);
	    }
	}

    [IncorrectImplementation]
    public class WordsStatisticsEN1 : IWordsStatistics
    {
        private IDictionary<string, int> statistics
            = new Dictionary<string, int>();

        public void AddWord(string word)
        {
            if (word == null) throw new ArgumentNullException(nameof(word));
            if (string.IsNullOrWhiteSpace(word)) return;
            if (word.Length > 10)
                word = word.Substring(0, 10);
            int count;
            statistics[word.ToLower()] = 1 + (statistics.TryGetValue(word.ToLower(), out count) ? count : 0);
        }

        public IEnumerable<WordCount> GetStatistics()
        {
            var temp = statistics;
            statistics = new Dictionary<string, int>();
            return temp
                .Select(WordCount.Create)
                .OrderByDescending(wordCount => wordCount.Count)
                .ThenBy(wordCount => wordCount.Word);
        }
    }

    [IncorrectImplementation]
    public class WordsStatisticsEN2 : WordsStatistics
    {
        private List<WordCount> result;

        public override IEnumerable<WordCount> GetStatistics()
        {
            return result ?? (result = base.GetStatistics().ToList());
        }
    }

    #endregion
}

File: js/src/challenge/incorrectImplementations/doNotOpen.js
import * as stringHelpers from "../infrastructure/stringHelpers";
import ArgumentNullError from "../infrastructure/argumentNullError";
import "../../lib/zip";

import WordsStatistics from "../wordsStatistics";

// <editor-fold defaultstate="collapsed" desc="Do not open!">

export class WordsStatisticsL2 extends WordsStatistics {
    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (stringHelpers.isWhitespace(word)) {
            return;
        }
        this.statistics.set(word.toLowerCase(), 1 + (this.statistics.get(word.toLowerCase()) || 0));
    }
}

export class WordsStatisticsL3 extends WordsStatistics {
    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (stringHelpers.isWhitespace(word)) {
            return;
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        } else if (word.length > 5) {
            word = word.substr(0, word.length - 2);
        }
        this.statistics.set(word.toLowerCase(), 1 + (this.statistics.get(word.toLowerCase()) || 0));
    }
}

export class WordsStatisticsL4 extends WordsStatistics {
    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (stringHelpers.isWhitespace(word)) {
            return;
        }
        if (word.length - 1 > 10) {
            word = word.substr(0, 10);
        }
        this.statistics.set(word.toLowerCase(), 1 + (this.statistics.get(word.toLowerCase()) || 0));
    }
}

export class WordsStatisticsC extends WordsStatistics {
    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (stringHelpers.isWhitespace(word)) {
            return;
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        if (!this.statistics.has(word.toLowerCase())) {
            this.statistics.set(word, 0);
        }
        this.statistics.set(word, 1 + this.statistics.get(word));
    }
}

export class WordsStatisticsE extends WordsStatistics {
    addWord(word) {
        if (!stringHelpers.isDefinedString(word) || stringHelpers.isWhitespace(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        this.statistics.set(word.toLowerCase(), 1 + (this.statistics.get(word.toLowerCase()) || 0));
    }
}

export class WordsStatisticsE2 extends WordsStatistics {
    addWord(word) {
        if (!stringHelpers.isDefinedString(word) || stringHelpers.isWhitespace(word)) {
            return;
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        this.statistics.set(word.toLowerCase(), 1 + (this.statistics.get(word.toLowerCase()) || 0));
    }
}

export class WordsStatisticsE3 extends WordsStatistics {
    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        if (stringHelpers.isWhitespace(word)) {
            return;
        }
        this.statistics.set(word.toLowerCase(), 1 + (this.statistics.get(word.toLowerCase()) || 0));
    }
}

export class WordsStatisticsE4 extends WordsStatistics {
    addWord(word) {
        if (word.length === 0 || stringHelpers.isWhitespace(word)) {
            return;
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        this.statistics.set(word.toLowerCase(), 1 + (this.statistics.get(word.toLowerCase()) || 0));
    }
}

export class WordsStatisticsO1 extends WordsStatistics {
    getStatistics() {
        return Array
            .from(this.statistics)
            .map((keyValue) => ({
                word: keyValue[0],
                count: keyValue[1]
            }))
            .sort((a, b) => a.word.localeCompare(b.word));
    }
}

export class WordsStatisticsO2 extends WordsStatistics {
    getStatistics() {
        return Array
            .from(this.statistics)
            .map((keyValue) => ({
                word: keyValue[0],
                count: keyValue[1]
            }))
            .sort((a, b) => b.count - a.count);
    }
}

export class WordsStatisticsO3 extends WordsStatistics {
    getStatistics() {
        return super.getStatistics().sort((a, b) => a.word.localeCompare(b.word));
    }
}

export class WordsStatisticsO4 extends WordsStatistics {
    getStatistics() {
        return super.getStatistics().sort((a, b) => a.count !== b.count
            ? b.count - a.count
            : -a.word.localeCompare(b.word));
    }
}

export class WordsStatisticsCR {
    constructor() {
        this.statistics = new Map();
    }

    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (stringHelpers.isEmpty(word)) {
            return;
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        word = word.toLowerCase();

        this.statistics.set(word, 1 + (this.statistics.get(word) || 0));
    }

    getStatistics() {
        return Array
            .from(this.statistics)
            .map((keyValue) => ({
                word: keyValue[0],
                count: keyValue[1]
            }))
            .sort((a, b) => a.count !== b.count
                ? b.count - a.count
                : a.word.localeCompare(b.word));
    }
}

const statistics_STA = new Map();

export class WordsStatisticsSTA {
    constructor() {
        statistics_STA.clear();
    }

    get statistics() {
        return statistics_STA;
    }

    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (stringHelpers.isWhitespace(word)) {
            return;
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        this.statistics.set(word.toLowerCase(), 1 + (this.statistics.get(word.toLowerCase()) || 0));
    }

    getStatistics() {
        return Array
            .from(this.statistics)
            .map((keyValue) => ({
                word: keyValue[0],
                count: keyValue[1]
            }))
            .sort((a, b) => a.count !== b.count
                ? b.count - a.count
                : a.word.localeCompare(b.word));
    }
}

export class WordsStatistics123 {
    constructor() {
        this.MAX_SIZE = 1237;

        this.statistics = new Array(this.MAX_SIZE);
        this.words = new Array(this.MAX_SIZE);
    }

    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (stringHelpers.isWhitespace(word)) {
            return;
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        const index = Math.abs(stringHelpers.calculateHash(word.toLowerCase())) % this.MAX_SIZE;
        this.statistics[index] = 1 + (this.statistics[index] || 0);
        this.words[index] = word.toLowerCase();
    }

    getStatistics() {
        return this.statistics
            .zip(this.words, (s, w) => ({ word: w, count: s }))
            .filter(s => s.count > 0)
            .sort((a, b) => a.count !== b.count
                ? b.count - a.count
                : a.word.localeCompare(b.word))
    }
}

export class WordsStatisticsQWE {
    constructor() {
        this.statistics = new Map();
    }

    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (stringHelpers.isWhitespace(word)) {
            return;
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        this.statistics.set(this.toLower(word), 1 + (this.statistics.get(this.toLower(word)) || 0));
    }

    getStatistics() {
        return Array
            .from(this.statistics)
            .map((keyValue) => ({
                word: keyValue[0],
                count: keyValue[1]
            }))
            .sort((a, b) => a.count !== b.count
                ? b.count - a.count
                : a.word.localeCompare(b.word));
    }

    toLower(word) {
        const englishCharsArray = [..."QWERTYUIOPLJKHGFDSAZXCVBNM"];
        const russianCharsArray = [..."ЙЦУКЕНГШЩЗФЫВАПРОЛДЯЧСМИТЬ"];

        const lowerCharacter = character => {
            const charCode = character.charCodeAt(0);
            if (englishCharsArray.includes(character)) {
                return String.fromCharCode(charCode - "D".charCodeAt(0) + "d".charCodeAt(0));
            } else if (russianCharsArray.includes(character)) {
                return String.fromCharCode(charCode - "Я".charCodeAt(0) + "я".charCodeAt(0));
            }
            return charCode;
        };

        return String.fromCharCode(...[...word].map(lowerCharacter));
    };
}

export class WordsStatistics998 {
    constructor() {
        this.statistics = [];
    }

    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (stringHelpers.isWhitespace(word)) {
            return;
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        let lowerCaseWord = word.toLowerCase();
        let wordCount = this.statistics.find(s => s.word === lowerCaseWord) || null;
        if (wordCount !== null) {
            this.statistics = this.statistics.filter(s => s.word !== wordCount.word || s.count !== wordCount.count);
        } else {
            wordCount = { word: lowerCaseWord, count: 0 };
        }
        this.statistics.push({ word: wordCount.word, count: wordCount.count - 1 });
        this.statistics.sort((a, b) => a.count === b.count
            ? a.word.localeCompare(b.word)
            : a.count - b.count);
    }

    getStatistics() {
        return this.statistics.map(s => ({word: s.word, count: -s.count}))
    }
}

export class WordsStatistics999 {
    constructor() {
        this.usedWords = new Set();
        this.statistics = [];
    }

    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (stringHelpers.isWhitespace(word)) {
            return;
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        word = word.toLowerCase();

        if (this.usedWords.has(word)) {
            let wordCount = this.statistics.find(s => s.word === word) || null;
            this.statistics = this.statistics.filter(s => s.word !== wordCount.word || s.count !== wordCount.count);
            this.statistics.push({ word: wordCount.word, count: wordCount.count + 1 });
        } else {
            this.statistics.push({ word: word, count: 1 });
            this.usedWords.add(word);
        }
    }

    getStatistics() {
        return this.statistics.sort((a, b) => a.count !== b.count
            ? b.count - a.count
            : a.word.localeCompare(b.word));
    }
}

export class WordsStatisticsEN1 {
    constructor() {
        this.statistics = new Map();
    }

    addWord(word) {
        if (!stringHelpers.isDefinedString(word)) {
            throw new ArgumentNullError("word should be a defined string");
        }
        if (stringHelpers.isWhitespace(word)) {
            return;
        }
        if (word.length > 10) {
            word = word.substr(0, 10);
        }
        this.statistics.set(word.toLowerCase(), 1 + (this.statistics.get(word.toLowerCase()) || 0));
    }

    getStatistics() {
        const temp = this.statistics;
        this.statistics = new Map();
        return Array
            .from(temp)
            .map((keyValue) => ({
                word: keyValue[0],
                count: keyValue[1]
            }))
            .sort((a, b) => a.count !== b.count
                ? b.count - a.count
                : a.word.localeCompare(b.word));
    }
}

export class WordsStatisticsEN2 extends WordsStatistics {
    constructor() {
        super();
        this.result = null;
    }

    getStatistics() {
        if (this.result) {
            return this.result;
        }
        this.result = super.getStatistics();
        return this.result;
    }
}

// </editor-fold>